### PR TITLE
Add ITTAPI hooks to jl_mutex_wait

### PR DESCRIPTION
**Pull request by pchintalapudi**
This allows identifying our `jl_mutex_t`s as mutexes by VTune

### CODE REVIEW

File: src/codegen.cpp [hunk 1]
```diff
@@ -8894,6 +8894,13 @@ extern "C" void jl_init_llvm(void)
     if (jit_profiling && atoi(jit_profiling)) {
         jl_using_intel_jitevents = 1;
     }
+    if (!jl_using_intel_jitevents) {
+        // Check if we are running under Intel VTune Amplifier
+        const char *vtune = getenv("INTEL_JIT_PROFILER64");
+        if (vtune && strlen(vtune) > 0) {
+            jl_using_intel_jitevents = 1;
+        }
+    }
```
Comment (vchuravy) [inline src/codegen.cpp:8903]: Move this into a separate PR? It would also need to detect `INTEL_JIT_PROFILER32` I think.

File: src/threading.c [hunk 2]
```diff
@@ -724,16 +728,61 @@ JL_DLLEXPORT void jl_exit_threaded_region(void)
...
+static void profile_lock_init(jl_mutex_t *lock, const char *name) JL_NOTSAFEPOINT {
+#ifdef USE_ITTAPI
+    //ITTAPI implementation
+    __itt_sync_create(lock, "jl_mutex_t", name, __itt_attr_mutex);
+#endif
+}
+
+static void profile_lock_start_wait(jl_mutex_t *lock) JL_NOTSAFEPOINT {
+#ifdef USE_ITTAPI
+    //ITTAPI implementation
```
Comment (vchuravy) [inline src/threading.c:744]: Uncesseary comment.

**Review by vchuravy — COMMENTED** [review]
@topolarity you might want to take a look at this for tracy integration

**Comment by vchuravy** [comment]
@pchintalapudi can you split this into two PRs?
This one for jl_mutex_wait, and a second one instead of hooking the probe points we should use ittapi as a timings backend like tracy #49140

File: Make.inc [hunk 3]
```diff
@@ -740,6 +740,12 @@ JCFLAGS += -DUSE_ITTAPI
 LIBITTAPI:=-littnotify
 endif
 
+ifeq ($(WITH_ITTAPI), 1)
+ifeq ($(WITH_DTRACE), 1)
+$(error "Only one of ITTAPI or DTRACE can be enabled")
+endif
+endif
+
```
Comment (vchuravy) [inline Make.inc:748]:
```suggestion
```

**Review by vchuravy — APPROVED** [review]
@pchintalapudi Can you also share some of the screenshots we looked at earlier?

File: src/threading.c [hunk 4]
```diff
@@ -724,16 +728,57 @@ JL_DLLEXPORT void jl_exit_threaded_region(void)
     }
 }
 
+// Profiling stubs
+
+#ifdef USE_ITTAPI
+
+#define profile_lock_init(lock, name) __itt_sync_create(lock, "jl_mutex_t", name, __itt_attr_mutex)
```
Comment (topolarity) [inline src/threading.c:735]: Would it be possible to move these definitions into `timing.c/.h`?

Ideally this would follow the conventions established there to provide backend-specific implementations (in particular, it should be possible to turn on multiple back-end implementations at once)

Comment (topolarity) [reply src/threading.c:735]: I'm also OK if these stay local to `threading.c`, but I would like to try to support multiple back-ends being enabled at once

**Comment by pchintalapudi** [comment]
The profiling events have been split into #49448.

**Comment by topolarity** [comment]
Thanks for this work! The bulk of the changes look great to me, although I did request a bit of re-factoring to better align with `jl_timing_*`

Nice to see some example data already too 🙂  Out of curiosity/ignorance, are any of these mutexes considerably "hot" or are they rare enough that we essentially don't need to worry about overhead?

P.S. Can you think of any significant locks that this change leaves out? I believe I/O has non-jl_mutex_t locks, or am I mistaken?

File: src/threading.c (hunk 4 above)
Comment (topolarity) [reply src/threading.c:735]: Would a static inline function fix that problem?

**Comment by pchintalapudi** [comment]
I/O and task switching have uv_mutex_t locks, but VTune at least picks those up automatically by tapping pthread_mutex_wait.

**Comment by pchintalapudi** [comment]
WRT hot locks, this workload was an intentionally bad one for the codegen lock, and I've seen the method's write lock get heavily contended when type inference runs in parallel trying to infer the same method with different arguments.

**Comment by topolarity** [comment]
I think I'm less worried about contention (which I'd want the profiler to show) and more worried about hot locks that are uncontended, where the profiler overhead might be significant relative to the lock / critical section

**Comment by pchintalapudi** [comment]
I can't think of a case where we acquire and release a lock in a hot loop, but if there was such a case I think it would be better to fix the acquire and release in the user code? We could also just insert a single trial of the cmpxchg that doesn't run the profiler to skip out on the uncontended case as well.

**Comment by topolarity** [comment]
>  We could also just insert a single trial of the cmpxchg that doesn't run the profiler to skip out on the uncontended case as well.

Yeah, that's a nice idea - I think there's a lot of effective tricks we could do to catch exceptional events for very hot locks.

If our existing locks aren't that hot though, let's just record all the events for now. If the overhead becomes suspect later, we can go back and try to use some tricks to reduce it.

File: src/threading.c (hunk 4 above)
Comment (vchuravy) [reply src/threading.c:735]: Yeah... Let's go with the static_inline

**Review by topolarity — APPROVED** [review]
All good on my end :+1:
